doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,616 @@
+
+By default the ``TypeResolver`` is implemented by searching all of the implementations
+listed in ``ufunc.types`` in order and stopping if all inputs can be safely
+cast to fit the signature.
+This means that if long (``l``) and double (``d``) arrays are added,
+numpy will find that the ``'dd->d'`` definition works
+(long can safely cast to double) and uses that.
+
+In some cases this is not desirable. For example the ``np.isnat`` universal
+function has a ``TypeResolver`` which rejects integer inputs instead of
+allowing them to be cast to float.
+In principle, downstream projects can currently use their own non-default
+``TypeResolver``, since the corresponding C-structure necessary to do this
+is public.
+The only project known to do this is Astropy, which is willing to switch to
+a new API if NumPy were to remove the possibility to replace the TypeResolver.
+
+For user defined datatypes, the dispatching logic is similar,
+although separately implemented and limited (see discussion below).
+
+
+Issues and Discussion
+"""""""""""""""""""""
+
+It is currently only possible for user defined functions to be found/resolved
+if any of the inputs (or the outputs) has the user datatype, since it uses the
+`OO->O` signature.
+For example, given that a ufunc loop to implement ``fraction_divide(int, int)
+-> Fraction`` has been implemented, 
+the call ``fraction_divide(4, 5)`` (with no specific output dtype) will fail
+because the loop that
+includes the user datatype ``Fraction`` (as output) can only be found if any of
+the inputs is already a ``Fraction``.
+``fraction_divide(4, 5, dtype=Fraction)`` can be made to work, but is inconvenient.
+
+Typically, dispatching is done by finding the first loop that matches. A match
+is defined as: all inputs (and possibly outputs) can
+be cast safely to the signature typechars (see also the current implementation
+section).
+However, in some cases safe casting is problematic and thus explicitly not
+allowed.
+For example the ``np.isnat`` function is currently only defined for
+datetime and timedelta,
+even though integers are defined to be safely castable to timedelta.
+If this was not the case, calling
+``np.isnat(np.array("NaT", "timedelta64").astype("int64"))`` would currently
+return true, although the integer input array has no notion of "not a time".
+If a universal function, such as most functions in ``scipy.special``, is only
+defined for ``float32`` and ``float64`` it will currently automatically
+cast a ``float16`` silently to ``float32`` (similarly for any integer input).
+This ensures successful execution, but allows a change in the output dtype
+when support for new data types is added to a ufunc.
+
+In general the order in which loops are registered is important.
+However, this is only reliable if all loops are added when the ufunc is first defined.
+Additional loops added when a new user datatypes is imported
+must not be sensitive to the order in which imports occur.
```

Comment:
I'm not sure I'm understanding this correctly. What I gather is that the ufunc type resolution is sensitive to the ordering of the loops during the type lookup, and this is another feature of the current dtype system that complicates type resolution for new, user-defined types.

Reply:
Basically, order having an affect is not desirable. A new, improved design (which is necessary, this being one of the issues) should fix that problem.